smolyay/test_function_class.py

```diff
@@ -0,0 +1,6478 @@
+import abc
+
+import numpy
+
+class test_fun(abc.ABC):
+    """Abstract test function class
+
+    Stores data about each test function including call, name, the number of
+    dimensions, the domain, and how to normalize that domain to the interval
+    [-1,1]
+
+    Parameters
+    ----------
+    name : str
+        the name of the class
+
+    dim : int
+        the number of variables used as input
+
+    lower_bounds : list of floats of size 1xdim
+        the lower bounds of the domain of each variable
+
+    upper_bounds : list of floats of size 1xdim
+        the upper bounds of the domain of each variable
+
+    domain : list of lists
+        the domain of the inputs for each dimension
+    """
+    def __call__(self,x):
+        """Evaluate the test function
+
+        Parameters
+        ----------
+        x : list of floats
+            the input to the call method
+
+        Raises
+        ------
+        ValueError
+            if the input is outside is outside the domain
+        """
+        if self.dim > 1:
+            oob = any(
+                xi < bound[0]-1e-10 or xi > bound[1]+1e-10 for xi,bound in
+                zip(x,self.domain))
+        else:
+            oob = x > self.upper_bounds[0] or x < self.lower_bounds[0]
+        if oob:
+            raise ValueError("Input out domain of function.")
+        return self._function(x)
+    
+    @property
+    def name(self):
+        """Name of the function"""
+        return self.__class__.__name__
+
+    @property
+    def dim(self):
+        """Number of variables"""
+        return len(self.domain)
+
+    @property
+    def lower_bounds(self):
+        """ the lower bounds of the domain of each variable"""
+        lower, _ = zip(*self.domain)
+        return list(lower)
```

Comment:
List comprehension is more direct here. Docstrings should for properties should also look like this:
```suggestion
        """list: Lower bound of each variable."""
        return [bound[0] for bound in self.domain]
```